### PR TITLE
Added shell for Eye-9

### PR DESCRIPTION
**Pull request by a-sidorova**
### Details:
 - *nGraph shell for Eye-9*
 - *Covered all exceptions by tests*
 - *[PR](https://github.com/openvinotoolkit/openvino/pull/11104) with specification*
 - *Updated spec: 3 inputs are required to avoid ctor conflicts*
 - *Fixed DetectionOutput conversion op tests - not initialized attribute `nms_threshold` (BTW, Valgrind shows this error in this test). [Failed tests are in CI](https://dev.azure.com/openvinoci/dldt/_build/results?buildId=319331&view=results)*

### Tickets:
 - *82301*

### CODE REVIEW

File: src/core/include/openvino/op/eyelike.hpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+namespace ov {
+namespace op {
+namespace v9 {
+/// \brief Tensor EyeLike operation.
+/// \ingroup ov_ops_cpp_api
+class OPENVINO_API EyeLike : public Op {
+public:
+    OPENVINO_OP("EyeLike", "opset9");
+    BWDCMP_RTTI_DECLARATION;
```
Comment (ilyachur) [inline src/core/include/openvino/op/eyelike.hpp:17]: You don't need to have this macro
```suggestion
```

File: src/core/src/op/eyelike.cpp [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+//
+
+#include "ngraph/op/eyelike.hpp"
+
+#include <ngraph/validation_util.hpp>
+
+#include "itt.hpp"
+
+using namespace std;
+using namespace ngraph;
```
Comment (ilyachur) [inline src/core/src/op/eyelike.cpp:12]:
```suggestion
```

File: src/core/src/op/eyelike.cpp [hunk 3]
```diff
@@ -0,0 +1,122 @@
+// Copyright (C) 2018-2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "ngraph/op/eyelike.hpp"
+
+#include <ngraph/validation_util.hpp>
```
Comment (ilyachur) [inline src/core/src/op/eyelike.cpp:7]: I think you can remove this header
```suggestion
```

File: src/core/src/op/eyelike.cpp [hunk 4]
```diff
@@ -0,0 +1,122 @@
+// Copyright (C) 2018-2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "ngraph/op/eyelike.hpp"
```
Comment (ilyachur) [inline src/core/src/op/eyelike.cpp:5]:
```suggestion
#include "openvino/op/eyelike.hpp"
```

File: src/core/include/openvino/op/eyelike.hpp (hunk 1 above)
Comment (a-sidorova) [reply src/core/include/openvino/op/eyelike.hpp:17]: Done, thanks

File: src/core/src/op/eyelike.cpp (hunk 3 above)
Comment (a-sidorova) [reply src/core/src/op/eyelike.cpp:7]: Unfortunately I cannot because I use `get_constant+from_source()`

File: src/core/src/op/eyelike.cpp (hunk 4 above)
Comment (a-sidorova) [reply src/core/src/op/eyelike.cpp:5]: Done, thanks

File: src/core/src/op/eyelike.cpp (hunk 2 above)
Comment (a-sidorova) [reply src/core/src/op/eyelike.cpp:12]: Done, thanks

**Comment by a-sidorova** [comment]
@jane-intel could you please take a look? Thanks!

File: src/core/src/op/eye.cpp [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+std::shared_ptr<ov::Node> ov::op::v9::Eye::clone_with_new_inputs(const ov::OutputVector& new_args) const {
+    NGRAPH_OP_SCOPE(v9_Eye_clone_with_new_inputs);
+    check_new_args_count(this, new_args);
+    if (new_args.size() == 1) {
+        return std::make_shared<v9::Eye>(new_args[0], m_output_type);
+    } else if (new_args.size() == 4) {
+        return std::make_shared<v9::Eye>(new_args[0], new_args[1], new_args[2], new_args[3], m_output_type);
+    } else {
+        throw std::runtime_error("Eye has incorrect input number: " + std::to_string(new_args.size()));
```
Comment (ilyachur) [inline src/core/src/op/eye.cpp:123]: Please use ov::Exception

**Review by ilyachur — APPROVED** [review]

File: src/core/src/op/eye.cpp (hunk 5 above)
Comment (a-sidorova) [reply src/core/src/op/eye.cpp]: Done, thanks!

File: src/core/shape_inference/include/eye_shape_inference.hpp [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+                    dims_batch_shape[i] = batch_shape[i];
+                }
+            }
+        } else {
+            output_shape = ov::PartialShape::dynamic();
+            return;
+        }
+    }
+
+    std::vector<ov::Dimension> output_dims(std::move(dims_batch_shape));
```
Comment (jane-intel) [inline src/core/shape_inference/include/eye_shape_inference.hpp:68]: there is no need to use vector -- PartialShape has all the overloads

File: src/core/shape_inference/include/eye_shape_inference.hpp [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    if (op->get_input_size() == 4) {
+        const auto batch_shape_pshape = input_shapes[3];
+        if (batch_shape_pshape.is_static()) {
+            dims_batch_shape.resize(batch_shape_pshape[0].get_length(), ov::Dimension::dynamic());
+            std::vector<int64_t> batch_shape;
+            if (get_data_as_int64<T>(3, op, batch_shape, constant_data)) {
+                NODE_VALIDATION_CHECK(op, batch_shape_pshape[0].get_length() == batch_shape.size());
+                dims_batch_shape.resize(batch_shape.size());
```
Comment (jane-intel) [inline src/core/shape_inference/include/eye_shape_inference.hpp:53]: why we do it twice? it was done on line 49

File: src/core/shape_inference/include/eye_shape_inference.hpp [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+                              num_columns.front());
+        dim_num_columns = num_columns.front();
+    }
+
+    if (op->get_input_size() == 4) {
+        const auto batch_shape_pshape = input_shapes[3];
+        if (batch_shape_pshape.is_static()) {
+            dims_batch_shape.resize(batch_shape_pshape[0].get_length(), ov::Dimension::dynamic());
+            std::vector<int64_t> batch_shape;
+            if (get_data_as_int64<T>(3, op, batch_shape, constant_data)) {
```
Comment (jane-intel) [inline src/core/shape_inference/include/eye_shape_inference.hpp:51]: Please consider dynamic values too -- look it up [here](https://github.com/openvinotoolkit/openvino/blob/4c659b76697883a409ddbcb4e088a97fa51d7202/src/core/shape_inference/include/broadcast_shape_inference.hpp#L192)

File: src/core/src/op/eye.cpp [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+    const auto& num_rows_pshape = get_input_partial_shape(0);
+    if (num_rows_pshape.is_static()) {
+        const auto& num_rows_rank = num_rows_pshape.rank().get_length();
+        NODE_VALIDATION_CHECK(this, num_rows_rank <= 1, "'num_rows' value must be a scalar or 1D tensor.");
+
+        if (num_rows_rank == 1) {
+            NODE_VALIDATION_CHECK(this, num_rows_pshape.compatible(ov::Shape{1}), "'num_rows' should have 1 element.");
+        }
+    }
```
Comment (jane-intel) [inline src/core/src/op/eye.cpp:45]:
```suggestion
```

File: src/core/src/op/eye.cpp [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+    const auto& num_columns_pshape = get_input_partial_shape(1);
+    if (num_columns_pshape.is_static()) {
+        const auto& num_columns_rank = num_columns_pshape.rank().get_length();
+        NODE_VALIDATION_CHECK(this, num_columns_rank <= 1, "'num_columns' value must be a scalar or 1D tensor.");
+        if (num_columns_rank == 1) {
+            NODE_VALIDATION_CHECK(this,
+                                  num_columns_pshape.compatible(ov::Shape{1}),
+                                  "'num_columns' should have 1 element.");
+        }
+    }
```
Comment (jane-intel) [inline src/core/src/op/eye.cpp:62]:
```suggestion
```

File: src/core/src/op/eye.cpp [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+    if (diagonal_index_pshape.is_static()) {
+        const auto& diagonal_index_rank = diagonal_index_pshape.rank().get_length();
+        NODE_VALIDATION_CHECK(this, diagonal_index_rank <= 1, "'diagonal_index' value must be a scalar or 1D tensor.");
+
+        if (diagonal_index_rank == 1) {
+            NODE_VALIDATION_CHECK(this,
+                                  diagonal_index_pshape.compatible(ov::Shape{1}),
+                                  "'diagonal_index' should have 1 element.");
+        }
+    }
```
Comment (jane-intel) [inline src/core/src/op/eye.cpp:79]:
```suggestion
```

File: src/core/src/op/eye.cpp [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+                              batch_shape_et);
+        const auto& batch_shape_pshape = get_input_partial_shape(3);
+        if (batch_shape_pshape.is_static()) {
+            const auto& diagonal_index_rank = batch_shape_pshape.rank().get_length();
+            NODE_VALIDATION_CHECK(this, diagonal_index_rank == 1, "'batch_shape' value must be a 1D tensor.");
+        } else {
+            NODE_VALIDATION_CHECK(this,
+                                  batch_shape_pshape.rank().is_static(),
+                                  "'batch_shape' should have static shape rank");
+            NODE_VALIDATION_CHECK(this, batch_shape_pshape.rank() == 1, "'batch_shape' value must be a 1D tensor.");
```
Comment (jane-intel) [inline src/core/src/op/eye.cpp:97]:
```suggestion
```

File: src/core/tests/type_prop/eye.cpp [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+
+#include "gtest/gtest.h"
+#include "ngraph/ngraph.hpp"
+#include "ngraph/opsets/opset9.hpp"
+#include "type_prop.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+TEST(type_prop, eye_constant) {
```
Comment (jane-intel) [inline src/core/tests/type_prop/eye.cpp:13]: Add a test where batch shape is being taken via `Parameter(PartialShape({1, 10}, {10, 25}) -> ShapeOf -(last input)-> Eye`
Output should contain `{1, 10}, {10, 25}` as batch.

File: src/core/tests/type_prop/eye.cpp [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+    auto num_rows = op::v0::Constant::create(element::i64, Shape{}, {6});
+    auto num_columns = op::v0::Constant::create(element::i64, Shape{}, {3});
+    auto diagonal_index = op::v0::Constant::create(element::i64, Shape{}, {0});
+
+    auto eye = std::make_shared<op::v9::Eye>(num_rows, num_columns, diagonal_index, element::bf16);
+
+    EXPECT_EQ(eye->get_output_element_type(0), element::bf16);
+    EXPECT_TRUE(eye->get_output_partial_shape(0).same_scheme(PartialShape{6, 3}));
```
Comment (jane-intel) [inline src/core/tests/type_prop/eye.cpp:21]: Please follow this approach in all these tests
```suggestion
    EXPECT_EQ(eye->get_output_partial_shape(0), PartialShape{6, 3});
```

File: src/tests/functional/inference_engine/transformations/op_conversions/detection_output_downgrade_test.cpp [hunk 15]
```diff
@@ -41,6 +41,7 @@ void create_attributes_vectors(std::vector<opset1::DetectionOutput::Attributes>&
                             variance_encoded_in_target;
                         attrs_v1.share_location = attrs_v8.share_location = share_location;
                         attrs_v1.normalized = attrs_v8.normalized = normalized;
+                        attrs_v1.nms_threshold = attrs_v8.nms_threshold = 0.5f;
```
Comment (jane-intel) [inline src/tests/functional/inference_engine/transformations/op_conversions/detection_output_downgrade_test.cpp:44]: ???

File: src/tests/functional/inference_engine/transformations/op_conversions/detection_output_upgrade_test.cpp [hunk 16]
```diff
@@ -39,6 +39,7 @@ void create_attributes_vectors(std::vector<opset1::DetectionOutput::Attributes>&
                             variance_encoded_in_target;
                         attrs_v1.share_location = attrs_v8.share_location = share_location;
                         attrs_v1.normalized = attrs_v8.normalized = normalized;
+                        attrs_v1.nms_threshold = attrs_v8.nms_threshold = 0.5f;
```
Comment (jane-intel) [inline src/tests/functional/inference_engine/transformations/op_conversions/detection_output_upgrade_test.cpp:42]: ???

**Review by jane-intel — CHANGES_REQUESTED** [review]

**Comment by jane-intel** [comment]
All shape-related checks should be contained in a template shape_infer function. Also, it should be introduced in shape_inference.cpp [like so](https://github.com/openvinotoolkit/openvino/blob/4c659b76697883a409ddbcb4e088a97fa51d7202/src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.cpp#L549)

File: src/core/src/op/eye.cpp (hunk 9 above)
Comment (a-sidorova) [reply src/core/src/op/eye.cpp]: Moved into shape_inference.cpp. Thanks!

File: src/core/src/op/eye.cpp (hunk 10 above)
Comment (a-sidorova) [reply src/core/src/op/eye.cpp]: Moved into shape_inference.cpp. Thanks!

File: src/core/src/op/eye.cpp (hunk 11 above)
Comment (a-sidorova) [reply src/core/src/op/eye.cpp]: Moved into shape_inference.cpp. Thanks!

File: src/core/src/op/eye.cpp (hunk 12 above)
Comment (a-sidorova) [reply src/core/src/op/eye.cpp]: Moved into shape_inference.cpp. Thanks!

File: src/core/tests/type_prop/eye.cpp (hunk 14 above)
Comment (a-sidorova) [reply src/core/tests/type_prop/eye.cpp]: Changed everywhere. Thanks

File: src/core/tests/type_prop/eye.cpp (hunk 13 above)
Comment (a-sidorova) [reply src/core/tests/type_prop/eye.cpp:13]: Cool test case! Added, thanks!

**Review by yury-intel — APPROVED** [review]

**Comment by a-sidorova** [comment]
@jane-intel all comments are applied.
Could you please review one more time?

**Review by tomdol — APPROVED** [review]

**Review by mitruska — APPROVED** [review]

**Pull request merged**